jm-alias/jm-alias.1.rst
########
jm-alias
########

Shell aliases
#############

:Manual section: 1
:Date: 2026-09-05
:Author: Jan Matějka jan@matejka.ninja
:Manual group: jm-util manual

SYNOPSIS
========

  jm alias [opts]

OPTIONS
=======

--show-aliases
  Print alias definitions.

--compdef <alias>
  Print compdef for <alias>

DESCRIPTION
===========

Shell aliases bad.

Commands good.

.. include:: ../core/common-foot.rst


debian/changelog
jm-util (0.4.0-1) unstable; urgency=medium

  [ Jan Matějka ]
  * ft(jm-claude)!: --account and --instance options

 -- Jan Matejka <jmatejka01@gmail.com>  Tue, 01 Sep 2026 23:07:00 +0200

jm-util (0.3.6-1) unstable; urgency=medium

  [ Jan Matějka ]
  * fx(completion): add missing commands and arguments
  * ft(oci): build on top of ghcr.io/jan-matejka/debian
  * ft(core/jm-tmux-autosession): keep org dir
  * fx(core/jm-claude#doc): args
  * ft(core/jm-container-env-setup): noop if not in git
  * fx(completion): add separator
  * rf: move jm-claude to its own module
  * ft(core/jm-claude): add --no-workdir
  * rf(jm-claude#doc): rename Howto -> USAGE
  * rm(oci): do not bake in sources
  * fx: master -> main; resistance is futile
  * fx(jm-claude#tests): rf: move jm-claude to its own module
  * ft(#build): add .dockerignore
  * ft(jm-claude): -p implies --no-workdir
  * ft: add jm-p-debug
  * fx(jm-claude#doc): rm incorrect SEE ALSO section
  * rm(#build): image target - broken and unused
  * rf(#build): rename image/service dev -> work
  * ft(#ci): add CI - build, push, & test
  * fx(git-cif): CI tests - check /dev/tty works before using it

 -- Jan Matejka <jmatejka01@gmail.com>  Sun, 23 Aug 2026 22:13:16 +0200

jm-util (0.3.5-1) unstable; urgency=medium

  [ Jan Matějka ]
  * ft: add jm-claude and switch to docker-compose under podman compose

 -- Jan Matejka <jmatejka01@gmail.com>  Sat, 22 Aug 2026 22:42:06 +0200

jm-util (0.3.4-1) unstable; urgency=medium

  [ Jan Matějka ]
  * ft(jm-alias): update aliases

 -- Jan Matejka <jmatejka01@gmail.com>  Sat, 15 Aug 2026 21:48:53 +0200

jm-util (0.3.3-1) unstable; urgency=medium

  [ Jan Matějka ]
  * ft: add git-lint
  * rf(oci): Containerfile multirefactor
  * fx: make packages - remove debian cargo from the image
  * fx(#build): cargo builds under debuild
  * fx(#build): cargo build with --release

 -- Jan Matejka <jmatejka01@gmail.com>  Sat, 15 Aug 2026 21:47:36 +0200

jm-util (0.3.2-1) unstable; urgency=medium

  [ Jan Matějka ]
  * fx(jm-alias): call by relative path
  * fx(jm-alias): pc completion

 -- Jan Matejka <jmatejka01@gmail.com>  Sat, 18 Jul 2026 17:05:22 +0200

jm-util (0.3.1-1) unstable; urgency=medium

  [ Jan Matějka ]
  * rf(versionator): simplify regex
  * fx(#build): make install-home - make i_bin_dir if missing
  * ft(#build): general support to build and install zsh completions
  * fx(jm-alias#build): make check runs dram_check
  * ft(oci#build): rewrite gitdirs only for debuild
  * fx(oci#build): overlays only for debuild
  * ft(oci/Containerfile#build): add libargparse-dev for jm-alias
  * ft(jm-alias): add --show-aliases
  * ft(jm-alias): add --compdef
  * ft(jm-alias): generate compdef files
  * fx(GNUmakefile#build): PHONY: packages mark
  * fx(versionator): always generate changelog (not just without -d)
  * fx(#build): building dev debian packages

 -- Jan Matejka <jmatejka01@gmail.com>  Sun, 28 Jun 2026 19:57:42 +0200

jm-util (0.3.0-1) unstable; urgency=medium

  [ Jan Matějka ]
  * fix(build): make image
  * fx: recurse make check
  * rf: add git-cif module
  * fx(git-cif): make git commit with un-edited template
  * ft(git-cif): git-cif is a superset of git-commit
  * fx(git-cif): handle case when lcpp is empty and run in single commit
  * rf(git-cif/jm-cif1.zsh): use -m instead of -F
  * fx(git-cif/jm-cif1.zsh): override EDITOR only if set to vim
  * ft(git-cif): -d does not open EDITOR by default
  * ft(jm-alias/jm-alias.cpp): grep exclude *.pyc
  * ft(jm-alias/jm-alias.cpp): grep --color=auto
  * fx(#build): dev build - key file name
  * fx(#build): install zsh_comp_dir
  * rf(#build): identify modules by file name .mod in the mod tree
  * ft(versionator): initial implementation
  * ft(#build): integrate versionator

 -- Jan Matejka <jmatejka01@gmail.com>  Sun, 28 Jun 2026 10:16:31 +0200

jm-util (0.2.1-1) unstable; urgency=medium

  [ Jan Matejka ]
  * Fix dependencies, tests.

 -- Jan Matejka <jmatejka01@gmail.com>  Wed, 26 Nov 2025 15:43:31 +0100

jm-util (0.2.0-1) unstable; urgency=medium

  [ Jan Matejka ]
  * Initial release.

 -- Jan Matejka <jmatejka01@gmail.com>  Tue, 21 May 2024 04:26:12 +0200


jm-alias/jm-alias.cpp
#include <unistd.h>

#include <algorithm>
#include <list>
#include <filesystem>
#include <format>
#include <iostream>
#include <map>
#include <optional>
#include <print>
#include <ranges>
#include <string>
#include <string_view>
#include <vector>

#include <argparse/argparse.hpp>

using namespace std;

struct AliasDef {
  const string alias;
  const string cmd;
  const vector<string> argv;
  const string compF;

  AliasDef(string alias, string cmd, vector<string> argv)
    : alias(move(alias)), cmd(std::move(cmd)), argv(std::move(argv)),
      compF(makeCompF(this->alias, this->cmd)) {}

  AliasDef(string alias, string cmd, vector<string> argv, string compF)
    : alias(move(alias)), cmd(std::move(cmd)), argv(std::move(argv)), compF(std::move(compF)) {}

private:
    static string makeCompF(const string& alias, const string& cmd) {
      constexpr auto fmt = R"EOF(#compdef {0}
_{0}() {{
  words[1]={1}
  service={1}
  _{1}
}}

_{0}
)EOF";
      return vformat(fmt, make_format_args(alias, cmd));
    }
};

const vector<AliasDef> v_aliases {
  AliasDef{"b", "buildah", {}},
  AliasDef{"g", "git", {}},
  AliasDef{"gr", "grep", {}},
  AliasDef{"d", "docker", {}},
  AliasDef{"dc", "docker-compose", {}},
  AliasDef{"l", "ls", {}},
  AliasDef{"ll", "ls", {"-l"}},
  AliasDef{"grr", "grep", {"-r", "--color=auto", "--exclude=*.pyc", "--exclude", "tags", "--exclude-dir=.git", "--exclude-dir=.tox"}},
  AliasDef{"p", "podman", {}},
  AliasDef{"dpc", "podman-compose", {}, "#compdef pc\ncomplete -F _podmanCompose dpc"},
  AliasDef{"pc", "jm", {"container-env-setup", "podman", "compose"}},
  // podman-compose completion needs to be set up from bash completion script in zshrc
  AliasDef{"s", "systemctl", {}},
  AliasDef{"t", "tmux", {}},
  AliasDef{"gr_video", "grep", {"-iE", "(avi|flv|mkv|wmv|mpg|mpeg|mp4)"}},
  AliasDef{"gr_pics", "grep", {"-iE", "(jpg|jpeg|tiff|bmp|png|gif)"}},
  AliasDef{"j", "journalctl", {}},
  AliasDef{"x", "xargs", {}},
};

map<string, AliasDef> aliases;

void to_lower_inplace(char& c) {
  c = (char)tolower(c);
}

bool boolish(char* x) {
  if(!x)
    return false;

  string s(x);
  for_each(s.begin(), s.end(), to_lower_inplace);
  return (s == "on" || s == "1" || s == "true");
}

class BaseLog {
public:
  template <typename... T>
  void verbose(format_string<T...> fmt, T&& ...args) {
    string msg = vformat(fmt.get(), make_format_args(args...));
    _verbose(msg);
  }

  virtual void _verbose(string msg) {
  }
};

class Log : public BaseLog {
public:
  virtual void _verbose(string msg) {
    cerr << msg << endl;
  }
};

BaseLog* _log = new BaseLog();

bool _which_check_path(
  filesystem::path p,
  optional<const filesystem::path> skip
) {
  _log->verbose("checking {}", string(p));

  if(!(exists(p) && (is_regular_file(p) || is_symlink(p)))) {
    _log->verbose("  failed existence check", string(p));
    return false;
  }

  using filesystem::perms;
  auto path_canon = filesystem::canonical(p);
  if (skip.has_value() and skip.value() == path_canon) {
    _log->verbose("  skip due to match with skip argument");
    return false;
  }

  auto ps = filesystem::status(p).permissions();
  if (!(
    perms::none != (ps & perms::owner_exec)
    || perms::none != (ps & perms::group_exec)
    || perms::none != (ps & perms::others_exec)
  )) {
    _log->verbose("  failed permissions passed");
    return false;
  }

  return true;
};

optional<string> which(
  string name,
  optional<const filesystem::path> skip
) {
  if (name.find("/") != string::npos) {
    filesystem::path p{name};
    if (_which_check_path(p, skip))
      return p;
  }

  char* c_path = getenv("PATH");
  if(!c_path)
    return nullopt;

  for(const auto& v_path: views::split(string(c_path), ':')) {
    filesystem::path p{string_view(v_path)};
    p /= name;
    if(_which_check_path(p, skip))
      return p;
  }

  return nullopt;
}

int ls_aliases() {
  for(const auto &[key, ignore]: aliases) {
    cout << format("{}\n", key);
  }
  return 0;
}

int dispatch(
  const string& name,
  vector<char*> args,
  const filesystem::path& self
) {
  if(!aliases.contains(name)) {
    println(cerr, "jm-alias: {}: alias not found", name);
    return 1;
  }

  auto target = aliases.at(name);
  auto path = which(target.cmd, self);
  if (!path) {
    println(cerr, "jm-alias: {}: command not found", target.cmd);
    return 1;
  }

  args.reserve(args.size() + target.argv.size());
  for(auto& x: target.argv | views::reverse) {
    args.insert((args.begin()+1), const_cast<char*>(x.c_str()));
  }
  args.push_back(nullptr);

  if(-1 == execv(path.value().c_str(), args.data()))
    perror("jm-alias: execv() failed");

  return 1;
}


class Cmd {
public:
  const filesystem::path cmd;

  Cmd(const filesystem::path cmd) : cmd(cmd) {}

  filesystem::path self() {
    auto with_path = which(cmd, nullopt);
    if (!with_path.has_value())
      throw runtime_error("jm-util: couldn't find self in PATH");

    const filesystem::path self = filesystem::canonical(with_path.value());
    return self;
  }

  const string name() {
    return cmd.filename();
  }
};

const string join(vector<string> xs, const string &sep) {
  string r;
  for(auto &x: xs) {
    if (!r.empty()) r += sep;
    r += x;
  }
  return r;
}

int show() {
  for(const auto &[name, alias]: aliases) {
    auto args = join(alias.argv, " ");
    cout << format("{} _{} {} {}\n", name, alias.cmd, alias.cmd, args);
  }

  return 0;
}

int compdef(const string &alias) {
  if(!aliases.contains(alias)) {
    println(cerr, "jm-alias: {}: alias not found", alias);
    return 1;
  }

  auto target = aliases.at(alias);
  println("{}", target.compF);
  return 0;
}

int main(const int argc, char** argv) {
  for (const auto& a : v_aliases) {
    aliases.emplace(a.alias, a);
  }

  if(boolish(getenv("JMU_ALIAS_VERBOSE")))
    _log = new Log();

  vector<char*> args(argv, argv+argc);

  auto cmd = Cmd(args[0]);

  if (cmd.name() == "jm-alias") {
    argparse::ArgumentParser args("jm-alias");

    args.add_argument("--show-aliases")
        .help("show aliases")
        .default_value(false)
        .implicit_value(true);

    args.add_argument("--compdef")
        .nargs(1);
    try {
      args.parse_args(argc, argv);
    } catch (const exception &err) {
      cerr << err.what() << endl;
      cerr << args;
      return 1;
    }
    if (args.is_used("--show-aliases"))
      return show();

    if (args.is_used("--compdef"))
      return compdef(args.get<string>("--compdef"));

    return ls_aliases();
  } else
    return dispatch(cmd.name(), args, cmd.self());
}
